sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala

```diff
@@ -0,0 +1,205 @@
+  extends RRunner[ColumnarBatch](
+    func,
+    "arrow",
+    "arrow",
+    packageNames,
+    broadcastVars,
+    numPartitions = -1,
+    isDataFrame = true,
+    schema.fieldNames,
+    RRunnerModes.DATAFRAME_GAPPLY) {
+
+  protected override def writeData(
+      dataOut: DataOutputStream,
+      printOut: PrintStream,
+      iter: Iterator[_]): Unit = if (iter.hasNext) {
+    val inputIterator = iter.asInstanceOf[Iterator[(Array[Byte], Iterator[InternalRow])]]
+    val arrowSchema = ArrowUtils.toArrowSchema(schema, timeZoneId)
+    val allocator = ArrowUtils.rootAllocator.newChildAllocator(
+      "stdout writer for R", 0, Long.MaxValue)
+    val root = VectorSchemaRoot.create(arrowSchema, allocator)
+    val out = new ByteArrayOutputStream()
+    val keys = collection.mutable.ArrayBuffer.empty[Array[Byte]]
+
+    Utils.tryWithSafeFinally {
+      val arrowWriter = ArrowWriter.create(root)
+      val writer = new ArrowStreamWriter(root, null, Channels.newChannel(out))
+      writer.start()
+
+      while (inputIterator.hasNext) {
+        val (key, nextBatch) = inputIterator.next()
+        keys.append(key)
+
+        while (nextBatch.hasNext) {
+          arrowWriter.write(nextBatch.next())
+        }
+
+        arrowWriter.finish()
+        writer.writeBatch()
+        arrowWriter.reset()
+      }
+      writer.end()
+    } {
+      // Don't close root and allocator in TaskCompletionListener to prevent
+      // a race condition. See `ArrowPythonRunner`.
+      root.close()
+      allocator.close()
+    }
+
+    // Currently, there looks no way to read batch by batch by socket connection in R side,
+    // See ARROW-4512. Therefore, it writes the whole Arrow streaming-formatted binary at
+    // once for now.
+    val data = out.toByteArray
+    dataOut.writeInt(data.length)
+    dataOut.write(data)
+
+    keys.foreach(dataOut.write)
+  }
+
+  protected override def newReaderIterator(
+      dataStream: DataInputStream, errThread: BufferedStreamThread): Iterator[ColumnarBatch] = {
+    new Iterator[ColumnarBatch] {
+      private val allocator = ArrowUtils.rootAllocator.newChildAllocator(
+        "stdin reader for R", 0, Long.MaxValue)
+
+      private var reader: ArrowStreamReader = _
+      private var root: VectorSchemaRoot = _
+      private var vectors: Array[ColumnVector] = _
+
+      TaskContext.get().addTaskCompletionListener[Unit] { _ =>
+        if (reader != null) {
+          reader.close(false)
+        }
+        allocator.close()
+      }
+
+      private var batchLoaded = true
+      private var nextObj: ColumnarBatch = _
+      private var eos = false
+
+      override def hasNext: Boolean = nextObj != null || {
+        if (!eos) {
+          nextObj = read()
+          hasNext
+        } else {
+          false
+        }
+      }
+
+      override def next(): ColumnarBatch = {
+        if (hasNext) {
+          val obj = nextObj
+          nextObj = null.asInstanceOf[ColumnarBatch]
+          obj
```

Comment:
this and hasNext is a bit weird, but I assume there's a reason for it..

Reply:
I agree the codes here looks a bit messy .. I mimicked the code structure from `PythonRunner`s for a possibility of deduplicating later. Should be okay.